Enable highlighting on title/summary/fullContent

marko-bekhta opened the pull request:
fixes https://github.com/quarkusio/search.quarkus.io/issues/39

I've used the unified highlighter as that is the one that Elasticsearch is maintaining on their own and have their own custom version. 
I had to remove `fullContent` field from excluded as otherwise we wouldn't be able to do highlighting on that field. Aaaand that lead to the problem as the content is really long and the search query will fail. I've increased the `max_analyzed_offset` from `1000000` to `5000000`. Alternatively, we could've use the `max_analyzed_offset` (https://www.elastic.co/guide/en/elasticsearch/reference/current/highlighting.html#max-analyzed-offset) but we didn't add it as an exposed option 🙈 and I didn't want to get into customizing the JSON 😃.

A couple of search results:

looking for a config property `quarkus.hibernate-search-orm.enabled`:
```json
{
  "total": 2,
  "hits": [
    {
      "id": "/guides/hibernate-search-orm-elasticsearch",
      "title": "Hibernate Search guide",
      "summary": [
        "Hibernate Search allows you to index your entities in an Elasticsearch cluster and easily offer full text search in all your Hibernate ORM-based applications."
      ],
      "content": [
        "at runtime Configuration property Type Default <span class=\"highlightedContent\">quarkus.hibernate</span>-<span class=\"highlightedContent\">search</span>-<span class=\"highlightedContent\">orm.enabled</span> Whether Hibernate <span class=\"highlightedContent\">Search</span> is enabled during the build.",
        "<span class=\"highlightedContent\">quarkus.hibernate</span>-<span class=\"highlightedContent\">search</span>-orm.\"",
        "<span class=\"highlightedContent\">quarkus.hibernate</span>-<span class=\"highlightedContent\">search</span>-orm.\"",
        "<span class=\"highlightedContent\">quarkus.hibernate</span>-<span class=\"highlightedContent\">search</span>-orm.\"",
        "<span class=\"highlightedContent\">quarkus.hibernate</span>-<span class=\"highlightedContent\">search</span>-orm.\""
      ]
    },
    {
      "id": "/guides/all-config",
      "title": "All configuration options",
      "summary": [
        "List all the configuration properties per extensions"
      ],
      "content": [
        "<span class=\"highlightedContent\">quarkus.hibernate</span>-<span class=\"highlightedContent\">search</span>-orm.\"",
        "<span class=\"highlightedContent\">quarkus.hibernate</span>-<span class=\"highlightedContent\">search</span>-orm.\"",
        "<span class=\"highlightedContent\">quarkus.hibernate</span>-<span class=\"highlightedContent\">search</span>-orm.\"",
        "<span class=\"highlightedContent\">quarkus.hibernate</span>-<span class=\"highlightedContent\">search</span>-orm.\"",
        "<span class=\"highlightedContent\">quarkus.hibernate</span>-<span class=\"highlightedContent\">search</span>-orm.\""
      ]
    }
  ]
}
```
looking for `hibernate search` :
```json
{
  "total": 32,
  "hits": [
    {
      "id": "/guides/hibernate-search-orm-elasticsearch",
      "title": "<span class=\"highlightedTitle\">Hibernate</span> <span class=\"highlightedTitle\">Search</span> guide",
      "summary": [
        "<span class=\"highlightedContent\">Hibernate</span> <span class=\"highlightedContent\">Search</span> allows you to index your entities in an Elasticsearch cluster and easily offer full text <span class=\"highlightedContent\">search</span> in all your <span class=\"highlightedContent\">Hibernate</span> ORM-based applications."
      ],
      "content": [
        "Will default to false in <span class=\"highlightedContent\">Hibernate</span> <span class=\"highlightedContent\">Search</span> 7.",
        "<span class=\"highlightedContent\">Search</span> guide You have a <span class=\"highlightedContent\">Hibernate</span> ORM-based application?",
        "If <span class=\"highlightedContent\">Hibernate</span> <span class=\"highlightedContent\">Search</span> is disabled during the build, all processing related to <span class=\"highlightedContent\">Hibernate</span> <span class=\"highlightedContent\">Search</span> will be skipped, but it will not be possible to activate <span class=\"highlightedContent\">Hibernate</span>",
        "Let’s see how you can do it with <span class=\"highlightedContent\">Hibernate</span> <span class=\"highlightedContent\">Search</span>.",
        "\\\n    -Dextensions='<span class=\"highlightedContent\">hibernate</span>-orm-panache,jdbc-postgresql,<span class=\"highlightedContent\">hibernate</span>-<span class=\"highlightedContent\">search</span>-orm-elasticsearch,resteasy-reactive-jackson' \\\n    -DnoCode\ncd <span class=\"highlightedContent\">hibernate</span>-<span class=\"highlightedContent\">search</span>-orm-elasticsearch-quickstart"
      ]
    },
    {
      "id": "/guides/hibernate-orm-panache",
      "title": "Simplified <span class=\"highlightedTitle\">Hibernate</span> ORM with Panache",
      "summary": [
        "<span class=\"highlightedContent\">Hibernate</span> ORM is the de facto Jakarta Persistence implementation and offers you the full breadth of an Object Relational Mapper."
      ],
      "content": [
        "-- <span class=\"highlightedContent\">Hibernate</span> ORM specific dependencies -->\n<dependency>\n    <groupId>io.quarkus</groupId>\n    <artifactId>quarkus-<span class=\"highlightedContent\">hibernate</span>-orm-panache</artifactId>\n</",
        "The solution is located in the <span class=\"highlightedContent\">hibernate</span>-orm-panache-quickstart directory.",
        "<span class=\"highlightedContent\">Hibernate</span> ORM with Panache focuses on making your entities trivial and fun to write in Quarkus.",
        "That’s all there is to it: with Panache, <span class=\"highlightedContent\">Hibernate</span> ORM has never looked so trim and neat.",
        "<span class=\"highlightedContent\">Hibernate</span> with Panache also allows for the use of the more classical repository pattern via PanacheRepository."
      ]
    }
  ]
}
```

yrodiere commented on src/main/java/io/quarkus/search/app/SearchService.java:
```diff
@@ -68,6 +69,21 @@ public SearchResult<SearchHit> search(@RestQuery @DefaultValue(QuarkusVersions.L
                 })
+                // * Highlighters are going to use spans-with-classes so that we will have more control over styling the visual on the search results screen.
+                // * The idea is that we want to get ~5 sentences (we use a sentence boundary scanner and set the fragment size to 0)
+                //   If we decide that we want to limit that and return only parts of a sentence (if those are really long ... we limit things by setting some positive value as a fragment size.
```
I think we do want smaller snippets.

See the UI: https://quarkus.io/guides/
We'll insert the snippets inside each card, so there's not much room, even if we replace the summary.

BTW, maybe the snippet size/count should be a parameter in the REST service? I'd have it default to count=0, meaning no highlighting at all.

marko-bekhta replied:
Yeah, that might be a nice option 😃 will do. I was thinking that we probably want to show the title and entire summary "no-matter-where-is-the-match" and then have a small section under with the highlights on content (small font and colors that stand out less). Let me give it a try with the UI and see 🙈

yrodiere replied:
> Yeah, that might be a nice option 😃 will do. I was thinking that we probably want to show the title and entire summary "no-matter-where-is-the-match" and then have a small section under with the highlights on content (small font and colors that stand out less).

+1, makes sense

Just FYI summaries are currently wrong in some documents, see #44 . Please ignore that problem, we'll solve it soon :)

> Let me give it a try with the UI and see 🙈

Thanks!

yrodiere replied:
Ah, and you might want to make the highlight CSS class a parameter as well... 

Maybe enable highlighting only when that parameter is set, and have the snippet size/count parameters default to something low?

yrodiere reviewed (commented):
Thanks.

If you have some time, can you also send a PR to my branch for the UI side? https://github.com/quarkusio/quarkusio.github.io/pull/1825
I guess we'd just add the highlights within the card of each result, i.e. here:

```
        <div v-for="card in cardHits"
            v-html="card.innerHTML"
            class="grid__item docs-card"
            :class="Array.from(card.classList)">
        </div>
```

I guess we'll need something like `v-html="card.innerHTML + <the highlighted snippets>"`

> I had to remove fullContent field from excluded as otherwise we wouldn't be able to do highlighting on that field. Aaaand that lead to the problem as the content is really long and the search query will fail. I've increased the max_analyzed_offset from 1000000 to 5000000.

Seems fair. Though... would there be a way to make Elasticsearch ignore content beyond the max offset, rather than failing?

> Alternatively, we could've use the max_analyzed_offset (https://www.elastic.co/guide/en/elasticsearch/reference/current/highlighting.html#max-analyzed-offset) but we didn't add it as an exposed option 🙈 and I didn't want to get into customizing the JSON 😃.

Well no you couldn't have, since the index setting takes precedence :) :

> The [max_analyzed_offset](https://www.elastic.co/guide/en/elasticsearch/reference/current/highlighting.html#max-analyzed-offset) query setting does not override the [index.highlight.max_analyzed_offset](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html#index-max-analyzed-offset) which prevails when it’s set to lower value than the query setting

yrodiere commented on src/main/java/io/quarkus/search/app/SearchService.java:
```diff
@@ -68,6 +69,21 @@ public SearchResult<SearchHit> search(@RestQuery @DefaultValue(QuarkusVersions.L
+                //   If we decide that we want to limit that and return only parts of a sentence (if those are really long ... we limit things by setting some positive value as a fragment size.
+                // * No match size is there to make sure that we are still going to get the text even if the field didn't have a match in it.
+                // * The title in the Guide entity is `Length.LONG` long, so we use that as a max value for no-match size, but hopefully nobody writes a title that long...
+                .highlighter(
+                        f -> f.unified().noMatchSize(Length.LONG).numberOfFragments(5).fragmentSize(0).orderByScore(true)
+                                .tag("<span class=\"highlightedContent\">", "</span>")
+                                .boundaryScanner().sentence().end())
+                // If we'd want to highlight a title in some way, e.g. make the text bolder, or add some background tint:
+                .highlighter("highlighter_title",
+                        f -> f.unified().numberOfFragments(1).tag("<span class=\"highlightedTitle\">", "</span>"))
+                // * If there's no match in the full content we don't want to return anything.
+                // * Also content is really huge, so we want to only get small parts of the sentences (we are setting a fragment size)
+                .highlighter("highlighter_content", f -> f.unified().noMatchSize(0).fragmentSize(150))
```
Also, can you enable HTML escaping? I see a few XML tags included in the output, and this could get dangerous.

marko-bekhta commented:
> Seems fair. Though... would there be a way to make Elasticsearch ignore content beyond the max offset, rather than failing?

right, so that's what that setting is supposed to do. The highlighter should stop looking further if the query setting value is reached, or am I misinterpreting the meaning of  that query value🙈

yrodiere commented:
> > Seems fair. Though... would there be a way to make Elasticsearch ignore content beyond the max offset, rather than failing?
> 
> right, so that's what that setting is supposed to do. The highlighter should stop looking further if the query setting value is reached, or am I misinterpreting the meaning of that query value🙈

Ah, ok, I thought it just made Elasticsearch fail sooner, which I guess was a dumb idea xD

Do you remember why we didn't include this in the API in the end? Might be worth creating a JIRA, even just to add comments explaining why we close it immediately as "Won't do" :D

marko-bekhta commented:
> Do you remember why we didn't include this in the API in the end? Might be worth creating a JIRA, even just to add comments explaining why we close it immediately as "Won't do" :D

It actually might not be that hard to add it ...  https://github.com/hibernate/hibernate-search/pull/3490 😅 I think it wasn't working as we were expecting, or not working for all highlighters and we decided to remove it and wait if there's a demand for it?

yrodiere commented on src/main/java/io/quarkus/search/app/SearchService.java:
```diff
@@ -68,6 +73,22 @@ public SearchResult<SearchHit> search(@RestQuery @DefaultValue(QuarkusVersions.L
+                // * If there's no match in the full content we don't want to return anything.
+                // * Also content is really huge, so we want to only get small parts of the sentences. We are allowing caller to pick the number of sentences and their length:
+                .highlighter("highlighter_content",
+                        f -> f.unified().noMatchSize(0).numberOfFragments(contentSnippets).fragmentSize(contentSnippetsLength))
```
I'm surprise this doesn't fail hard with the default values of 0?

marko-bekhta replied:
Highlights work in mysterious ways 😄 
I've changed the defaults a bit... I was expecting that with 0 `numberOfFragments`  it won't return anything, but turns out it returns all the content 🙈 
If we'd wanted to not return that part at all, then we'd need to have a different projection and pass it in conditionally... Do you think it's worth having that?

yrodiere replied:
> Do you think it's worth having that?

Not necessarily, but I think it's worth having defaults that work properly (e.g. `numberOfFragments = 5`) and to throw a proper exception for invalid input (feel free to add Hibernate Validator to the project to validate method parameters).

yrodiere replied:
> Not necessarily, but I think it's worth having defaults that work properly (e.g. numberOfFragments = 5)

Ok, seems you did that part. LGTM.

> and to throw a proper exception for invalid input (feel free to add Hibernate Validator to the project to validate method parameters).

This is not really necessary, I was getting ahead of myself. Please disregard.

yrodiere reviewed (commented):
Please just add a few tests to SearchServiceTest, and I think we're good to go :)

marko-bekhta commented:
I've added a couple of tests and a guide type; That should cover all the needs of the site. We don't really use keywords or categories; they are just written to some "invisible divs" to be used for JS search. So no need in other properties being returned, right?

While making these changes, I was looking at where we can get the external URL ... and I think we are missing those, no?
I see that there are some guides from `quarkiverse-hub` in https://github.com/quarkusio/quarkusio.github.io/blob/develop/_data/guides-2-7.yaml but they do not have local files, right? So I was wondering if we should also parse these files and append the missing `quarkiverse-hub` guides?

yrodiere commented:
> While making these changes, I was looking at where we can get the external URL ... and I think we are missing those, no?
> I see that there are some guides from `quarkiverse-hub` in https://github.com/quarkusio/quarkusio.github.io/blob/develop/_data/guides-2-7.yaml but they do not have local files, right? So I was wondering if we should also parse these files and append the missing `quarkiverse-hub` guides?

Ok, then I guess `url` actually *is* an URL, and I was wrong :) Then it's worth renaming the `Guide#path` attribute to `url`, and maybe changing its type from `String` to `URI` (I don't know if that works for relative URLs though...)

yrodiere commented on src/main/java/io/quarkus/search/app/dto/SearchHit.java:
```diff
@@ -1,9 +1,22 @@
 
-@ProjectionConstructor
-public record SearchHit(@IdProjection String id, @FieldProjection String title) {
+public record SearchHit(String path, String type, String title, String summary, Set<String> content) {
```
Considering this is now all very specific to Guides... maybe let's rename this to `GuideSearchHit`? And move the REST API endpoint to `/guides/search`?

marko-bekhta replied:
makes sense 😃

yrodiere commented:
LGTM overall, though I suggested a few cosmetic changes. Do you want to merge and work on these changes later, or apply them before merging?

yrodiere commented:
> > While making these changes, I was looking at where we can get the external URL ... and I think we are missing those, no?
> > I see that there are some guides from `quarkiverse-hub` in https://github.com/quarkusio/quarkusio.github.io/blob/develop/_data/guides-2-7.yaml but they do not have local files, right? So I was wondering if we should also parse these files and append the missing `quarkiverse-hub` guides?
> 
> Ok, then I guess `url` actually _is_ an URL, and I was wrong :) Then it's worth renaming the `Guide#path` attribute to `url`, and maybe changing its type from `String` to `URI` (I don't know if that works for relative URLs though...)

To clarify, the `quarkus.yaml` *does* include the full URL to those guides in the `url` field, which we use to initialize the `path` field in our model. So I was saying we already have the data you want, we just need to rename `path` to `url`.

marko-bekhta commented:
I'll make the changes before merging 😃

> To clarify, the quarkus.yaml does include the full URL to those guides in the url field, which we use to initialize the path field in our model. So I was saying we already have the data you want, we just need to rename path to URL.

I was thinking that since we are going through the "local" directories in `QuarkusIO#guideDirectories()` we never created an entity for a quarkiverse guide, or those are also in that directory (I haven't looked closely 🙈)
and for older versions we should take that info from the other yaml ...

yrodiere commented:
> > To clarify, the quarkus.yaml does include the full URL to those guides in the url field, which we use to initialize the path field in our model. So I was saying we already have the data you want, we just need to rename path to URL.
> 
> I was thinking that since we are going through the "local" directories in `QuarkusIO#guideDirectories()` we never created an entity for a quarkiverse guide, or those are also in that directory (I haven't looked closely 🙈) and for older versions we should take that info from the other yaml ...

I created #54 to address indexing of the full content of those guides.

Are you saying that we're even skipping indexing of *metadata* for the Quarkiverse guides?
If so, I agree we should fix that. It seems like an issue unrelated to this PR though, so perhaps create a separate ticket and/or update #54?

marko-bekhta commented:
yeah at least it seems so ...  let me address your suggestions and then I'll take a closer look in the ticket you've created 😃

marko-bekhta commented:
Relative urls worked OK, so I switched the type. I've also added a couple `@ElementCollection` to the entity so that IDE stop complaining about those sets of strings.

yrodiere commented:
> I've also added a couple `@ElementCollection` to the entity so that IDE stop complaining about those sets of strings.

The IDE was wrong :(

H2 supports SQL arrays just fine, so no `@ElementCollection` is needed:

* https://docs.jboss.org/hibernate/orm/6.2/userguide/html_single/Hibernate_User_Guide.html#collections-as-basic
* https://github.com/hibernate/hibernate-orm/blob/12442bd8c7cde6e7c006a6277eeb8e81ad0c2219/hibernate-core/src/main/java/org/hibernate/dialect/H2Dialect.java#L193

marko-bekhta commented:
aaand removed them 😄

yrodiere commented:
@marko-bekhta Thanks. I pushed a small fix. Let's merge once you adapted your [other PR](https://github.com/yrodiere/quarkusio.github.io/pull/1) to use the new REST path and new field names?

yrodiere commented:
Tests are failing because indexing fails...

Perhaps related, I see you didn't update the quarkusio sample included in the repo in your PR that changed how we parse metadata: see #52 , `QuarkusIOSample`. I suspect the sample included in test resources is no longer complete enough, because it's missing some `quarkus.yaml` files?

marko-bekhta commented:
> I suspect the sample included in test resources is no longer complete enough, because it's missing some quarkus.yaml files?

if there's no yaml it should've just fall back to parsing data from an asciidoc... but anyway, I've added these metadata files to the sample too. It changed the relevance tests as a result as now we get some of the summaries where they were missing from asciidoc parsing 😃. I've also mapped the guide types to what the quarkus.io is using as otherwise the icons next to the search results do not match what is displayed initially 🙈 
as for the failed tests ... I've seen something similar when I was restarting the service multiple times in quick succession, but what can I say.... the tests weren't failing for me when the rest endpoint wasn't defined either 😄

yrodiere commented on src/test/data/quarkusio/_data/versioned/latest/index/quarkus.yaml:
```diff
@@ -0,0 +1,2828 @@
+# Generated file. Do not edit
```
Did you include this on purpose? It's supposed to be in the sample ZIP already, no?

marko-bekhta replied:
yes, it's in the zip. but since there're a few of those ascidoc files present in that directory I was thinking that there should also be a copy of the yaml (for reasons unknown 😄)

yrodiere replied:
> since there're a few of those ascidoc files present in that directory

That's... not normal. I think I forgot to remove these.

I'll create a separate PR with your changes and a few more.

marko-bekhta replied:
ohh ok thanks 😃 
and sorry I've pushed a change removing that type-mapping and added `Awaitility.await().timeout(Duration.ofMinutes(5))` to see if that will help with the tests on GH since they work ok locally ...

yrodiere replied:
> I'll create a separate PR with your changes and a few more.

Done: https://github.com/quarkusio/search.quarkus.io/pull/55

yrodiere commented on src/test/data/quarkusio/_data/versioned/main/index/quarkus.yaml:
```diff
@@ -0,0 +1,2810 @@
+# Generated file. Do not edit
```
Did you include this on purpose? It's supposed to be in the sample ZIP already, no?

yrodiere commented on src/main/java/io/quarkus/search/app/yml/QuarkusMetadata.java:
```diff
@@ -18,6 +18,13 @@
 
 public class QuarkusMetadata {
 
+    private static final Map<String, String> TYPE_MAPPER = Map.of(
```
I'm sorry, what is this?

yrodiere replied:
Ah, that's whay your explained here, I guess:

> I've also mapped the guide types to what the quarkus.io is using as otherwise the icons next to the search results do not match what is displayed initially 🙈

Still, I don't understand why this is needed... I'd expect our data to be authoritative, and if there's some transformation going on to turn it into a class name, well... it should happen on the presentation side, e.g. in the vue.js app or LitElement webcomponent.

marko-bekhta replied:
https://github.com/quarkusio/quarkusio.github.io/blob/bdff06f6bd1848dcf1b616ed14529f999e19246a/_includes/index-docs.html#L114-L123

^ see how a `type=guide` is a `class="referencebkg"` ... so this is to address this mismatch. Alternatively, we change it on the `quarkusio` side and use the classes that would actually map to the types

yrodiere replied:
> ^ see how a `type=guide` is a `class="referencebkg"` ... so this is to address this mismatch. Alternatively, we change it on the `quarkusio` side and use the classes that would actually map to the types

Yes, I would much rather have that. We shouldn't incorporate logic related to CSS classes in a rest service.

yrodiere commented on src/test/java/io/quarkus/search/app/SearchServiceTest.java:
```diff
@@ -42,18 +50,18 @@ protected int managementPort() {
     @BeforeAll
     void waitForIndexing() {
-        Awaitility.await().untilAsserted(() -> {
+        Awaitility.await().timeout(Duration.ofMinutes(5)).untilAsserted(() -> {
```
It shouldn't be that slow... We just unzip a git repository and create a local clone...

On the other hand the default is 15 seconds, which does seem a bit tight.

I'll add a lower timeout to https://github.com/quarkusio/search.quarkus.io/pull/55.

yrodiere commented:
Ok I rebased on main after merging #55, I think we should be good now. Let me know when your other PR is ready @marko-bekhta , and we'll merge it all at the same time.

yrodiere commented:
Looks like tests still fail... because nothing gets indexed?

```
Checking out files:     100% (22/22)
2023-11-17 13:33:17,463 INFO  [io.qua.sea.app.ind.IndexingService] (executor-thread-1) Indexing quarkus.io...
2023-11-17 13:33:17,477 INFO  [io.qua.sea.app.ind.IndexingService] (executor-thread-1) Refreshing indexes...
2023-11-17 13:33:17,492 INFO  [io.qua.sea.app.ind.Rollover] (executor-thread-1) Committing index rollover
...
2023-11-17 13:33:17,532 INFO  [io.qua.sea.app.ind.IndexingService] (executor-thread-1) Indexing success
```

I don't see the "Indexed X documents" logs... meaning nothing got indexed?

marko-bekhta commented:
hmm ok .. 
```
2023-11-17 16:35:23,692 INFO  [io.qua.sea.app.ind.IndexingService] (executor-thread-1) About to persist: Guide{url='/guides/stork-reference'}
2023-11-17 16:35:23,693 INFO  [io.qua.sea.app.ind.IndexingService] (executor-thread-1) About to persist: Guide{url='/guides/spring-data-jpa'}
2023-11-17 16:35:23,693 ERROR [io.qua.sea.app.ind.IndexingService] (executor-thread-1) failed to persist: jakarta.persistence.EntityExistsException: A different object with the same identifier value was already associated with the session : [io.quarkus.search.app.entity.Guide#/guides/spring-data-jpa]
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:123)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:162)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:168)
	at org.hibernate.internal.SessionImpl.firePersist(SessionImpl.java:761)
	at org.hibernate.internal.SessionImpl.persist(SessionImpl.java:739)
	at io.quarkus.hibernate.orm.runtime.session.TransactionScopedSession.persist(TransactionScopedSession.java:145)
	at org.hibernate.engine.spi.SessionLazyDelegator.persist(SessionLazyDelegator.java:275)
	at org.hibernate.Session_3a974b6a18ac399f675913d732c105426414d370_Synthetic_ClientProxy.persist(Unknown Source)
	at io.quarkus.search.app.indexing.IndexingService.batchIndex(IndexingService.java:193)
	at io.quarkus.search.app.indexing.IndexingService.indexQuarkusIo(IndexingService.java:180)
	at io.quarkus.search.app.indexing.IndexingService.indexAll(IndexingService.java:157)
	at io.quarkus.search.app.indexing.IndexingService.reindex(IndexingService.java:106)
	at io.quarkus.search.app.indexing.IndexingService_Subclass.reindex$$superforward(Unknown Source)
	at io.quarkus.search.app.indexing.IndexingService_Subclass$$function$$1.apply(Unknown Source)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:73)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:62)
	at io.quarkus.arc.impl.ActivateRequestContextInterceptor.invoke(ActivateRequestContextInterceptor.java:124)
	at io.quarkus.arc.impl.ActivateRequestContextInterceptor.aroundInvoke(ActivateRequestContextInterceptor.java:33)
	at io.quarkus.arc.impl.ActivateRequestContextInterceptor_Bean.intercept(Unknown Source)
	at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:42)
	at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:30)
	at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:27)
	at io.quarkus.search.app.indexing.IndexingService_Subclass.reindex(Unknown Source)
	at io.quarkus.search.app.indexing.IndexingService.lambda$indexOnStartup$3(IndexingService.java:73)
	at io.smallrye.context.impl.wrappers.SlowContextualSupplier.get(SlowContextualSupplier.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateFromItemSupplier.subscribe(UniCreateFromItemSupplier.java:28)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:587)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base@21.0.1/java.lang.Thread.runWith(Thread.java:1596)
	at java.base@21.0.1/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:832)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.posix.thread.PosixPlatformThreads.pthreadStartRoutine(PosixPlatformThreads.java:211)
```

And that's only happening in the native build; if I switch back to string URL all is back to normal (in the native mode as well)

yrodiere commented:
Merging, thanks!
